linux/src/async_socket_linux.c

```diff
@@ -274,6 +282,12 @@ static void internal_close(ASYNC_SOCKET_HANDLE async_socket)
         completion_port_remove(async_socket->completion_port, async_socket->socket_handle);
```

Comment:
>completion_port_remove [](http://example.com/codeflow?start=8&length=22)

Looking in the code, it seems that `event_complete_callback` is given to the completion port.

So line 282 is detaching the socket from the completion port. 

Shouldn't THAT take care of no ongoing calls rather than this layer?

I would expect that at the call of `completion_port_remove` there's NO WAY in which not all the calls are exhauster or there's still ongoing threads that have started and not returned yet to the completion port.

Reply:
detaching the socket from the completion port seems to be async and so the failure we're seeing is a race between line 292 and line 152

Reply:
I've changed the state of this item from "WontFix" to "Active" because I truly believe that the completion port bears some responsibility here. And maybe the fix lies both with the completion port AND within this file.

Reply:
The reason the event_complete_executing is added to protect the event_complete_callback function.  This is to ensure we don't close while the event_complete_callback is executing.